elastic/kibana
# PoC: Generate OAS spec from versioned router declarations

## Summary

PoC to explore whether we can generate good/high quality OAS spec from our new versioned router. The general idea is to bring the versioned router interface as inline as possible with OAS 3 so that we can map from router declarations as easily as possible.

## How to test

* See the `README.md` for local testing instructions.
* See the `run.js` file for how we can declare the route and generate the spec.

## What this PoC does


<details>
<summary>Given router declaration like</summary>

```ts
router
  .get({
    access: 'public',
    path: '/api/cool/endpoint/{id}',
  })
  .addVersion(
    {
      version: '2023-01-01',
      description: 'My cool route',
      validate: {
        request: {
          params: z.object({ id: z.string() }),
        },
        response: {
          200: {
            body: z.object({ foo: z.string().max(5) }),
          },
          400: {
            body: z.object({ error: z.string() }),
          },
        },
      },
    },
    (ctx, req, res) => res.ok()
  )
```

</details>

<details>

<summary>Generate OAS 3 like</summary>

```json
{
  "openapi": "3.0.0",
  "info": {
    "title": "PoC",
    "version": "1.0.0"
  },
  "servers": [
    {
      "url": "http://localhost:5601"
    }
  ],
  "paths": {
    "/api/cool/endpoint/{id}": {
      "get": {
        "responses": {
          "200": {
            "description": "No description",
            "content": {
              "application/json; Elastic-Api-Version=2023-01-01": {
                "schema": {
                  "type": "object",
                  "properties": {
                    "foo": {
                      "type": "string",
                      "maxLength": 5
                    }
                  },
                  "required": [
                    "foo"
                  ],
                  "additionalProperties": false
                }
              }
            }
          },
          "400": {
            "description": "No description",
            "content": {
              "application/json; Elastic-Api-Version=2023-01-01": {
                "schema": {
                  "type": "object",
                  "properties": {
                    "error": {
                      "type": "string"
                    }
                  },
                  "required": [
                    "error"
                  ],
                  "additionalProperties": false
                }
              }
            }
          }
        },
        "parameters": [
          {
            "name": "id",
            "in": "path",
            "required": true,
            "schema": {
              "type": "string"
            }
          }
        ]
      }
    }
  }
}
```

</details>

The script for doing this will cover

- [x] Request body
- [x] Responses
- [x] URL parameters
- [ ] Query parameters (still need to test)
- [ ] What else...?

Also makes adjustments to the versioned router interface as needed.

### Potential benefits

* We can have a single schema declaration for our request, response, parameter, query types that will stay insync across: runtime validations, OAS spec and TypeScript
* We need to get all of our public HTTP APIs onto a Elastic version which requires a refactor, if we plan the refactor carefully we can extract even more value from that effort

### Potential drawbacks

* It is considered more standard to write spec and generate code, not the other way around. The risk is that our spec is ultimately not that useful or lacking in certain aspects. The mitigation to this concern should be that our router interface will closely match OAS 3 and so we should be able to get the full benefits of OAS for consumers of the spec.

### Out of scope

* This PoC assumes that the problem of generating JSONSchema from runtime validation is "solved" using `zod-to-json-schema` and so `zod`, not `config-schema` which uses `joi`, was used for schema declarations. To migrate `config-schema` to `zod` was considered out of scope and requires a different PoC to derisk (incoming).
* Not all OAS 3 capabilities are covered. We excluded examples....
* Covering all edge cases: there are many edge cases that exist that this version of the script does not cover but it is assumed that we will be able to address them if the core concerns are met.

### Screenshots

<img width="1359" alt="Screenshot 2023-04-21 at 16 29 45" src="https://user-images.githubusercontent.com/8155004/233662129-0757d04a-76d5-4e0b-9238-26cc645f0412.png">
